juju/juju
# 2.6 Retrieve Unit Config From Model Cache

## Description of change

This patch updates the uniter API to retrieve unit charm config settings from the model cache.

It is the first of two parts, the second of which will alter configuration _watching_ to work from the cache. It is broken up to ease reviewing because most of this patch ensures that the uniter test suite continues to work with the _cachetest_ package functionality mixed in.

It is known that watching the DB for changes and retrieving them from the cache is not a desirable state - the second patch will follow as quickly as can be managed.

The functionality change is a handful of lines in _apiserver/facades/agent/uniter/uniter.go_.

## QA steps

- Bootstrap
- `juju deploy redis`.
- `juju config redis password=new-pass`.
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that the line "requirepass new-pass" is present.
- `juju config redis databases=4`.
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that the line "databases 4" is present.
- `export JUJU_DEV_FEATURE_FLAGS=generations`.
- `juju add-branch test-branch`.
- `juju track test-branch redis/0`.
- `juju config redis password=branch-pass --branch test-branch`.
- `juju config redis databases=8 --branch master` (the watcher is still watching settings in state, so we need to invoke it by changing master settings).
- `juju ssh 0 cat /etc/redis/redis-charm.conf` and check that:
  - the line "databases 8" is present;
  - the line "requirepass branch-pass" (from branch-based settings) is present.

## Documentation changes

None.

## Bug reference

N/A
